src/mpi/romio/adio/common/ad_write_str.c

```diff
@@ -280,8 +280,30 @@ void ADIOI_GEN_WriteStrided(ADIO_File fd, const void *buf, int count,
              * datatypes, instead of a count of bytes (which might overflow)
              * Other WriteContig calls in this path are operating on data
              * sieving buffer */
+
+            /* The ranks with a contiguous view need to lock in order
+             * to not step on their MPI_File_write() peers, if any peers have
+             * a non-contig view and are thus potentially in a sieving path.
+             *
+             * Note: I don't think I need to be conditional on fd->atomicity
+             * here, the peer ranks with their non-contiguous view differ in
+             * the location they take a lock depending on whether they have
+             * atomicity set, but they end up with a lock either way.
+             */
+            int tsize;
```

Comment:
The HDF5 test suite triggered a related problem, but what I was seeing was separate independent operations racing ahead of collective operations and getting in the way of the read-modify-write.  My solution was to add a collective pause at the end which can definitely hurt performance.  Look for the Allreduce around line 1033. 

Now obviously that synchronization wasn't enough to fix your test case.  But between your additional locking and my additional synchronization, there is probably a more sophisticated solution we could devise that fixes both.  Not suggesting we spend the next year working on that!  Let's get a fix in place now.

Reply:
Sounds good. I remember back when I first put in my extra lock I tried to make related testcases hoping to get other I/O to sneak in while sieving was in place, but I wasn't able to get a failure.  Maybe because of the collective operation you mentioned, or maybe because a lot of write paths end up back where I added the new lock/unlock above.